File: renderers/android/src/main/kotlin/dev/varn/gui/VarnActions.kt
package dev.varn.gui

import android.view.View
import android.view.inputmethod.InputMethodManager
import org.json.JSONObject

/** Performs the imperative action a ref asked for, refusing a name the renderer has no answer for. */
object VarnActions {
    fun perform(method: String, view: View, arguments: JSONObject, density: Float): Boolean = when (method) {
        "focus" -> {
            view.requestFocus()
            val manager = view.context.getSystemService(InputMethodManager::class.java)
            manager?.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT)
            true
        }

        "blur" -> {
            view.clearFocus()
            hideKeyboard(view)
            true
        }

        "scrollTo" -> {
            val surface = view as? VarnCollectionView
                ?: throw VarnRendererException("scrollTo needs a scrolling view")

            surface.scrollTo(
                (arguments.optDouble("x", 0.0) * density).toFloat(),
                (arguments.optDouble("y", 0.0) * density).toFloat(),
                arguments.optBoolean("animated", false),
            )

            true
        }

        "capturePhoto", "startRecording", "stopRecording" -> {
            val camera = view as? VarnCameraView ?: throw VarnRendererException("$method needs a camera")

            when (method) {
                "capturePhoto" -> camera.capturePhoto()
                "startRecording" -> camera.startRecording()
                else -> camera.stopRecording()
            }

            true
        }

        "play", "pause" -> {
            val video = view as? VarnVideoView ?: throw VarnRendererException("$method needs a video")

            if (method == "play") video.start() else video.pause()
            true
        }

        "seek" -> {
            if (!arguments.has("seconds")) {
                throw VarnRendererException("seek is asked for a number of seconds")
            }

            val seconds = arguments.optDouble("seconds", 0.0)

            when (view) {
                is VarnAudioView -> view.seek(seconds)
                is VarnVideoView -> view.seekTo((seconds * 1000).toInt())
                else -> throw VarnRendererException("seek needs a sound or a film")
            }

            true
        }

        "toggleMark", "setLink", "clearLink", "copy", "cut", "paste", "selectAll" -> {
            val editor = view as? VarnRichEditor ?: throw VarnRendererException("$method needs an editor")

            editor.requestFocus()

            when (method) {
                "toggleMark" -> editor.toggle(arguments.optString("mark"))
                "setLink" -> editor.link(arguments.optString("url"))
                "clearLink" -> editor.link(null)
                else -> editor.clipboard(method)
            }

            true
        }

        else -> throw VarnRendererException("the renderer has no action named $method")
    }

    /** Takes the keyboard away, whatever it was brought up by. */
    fun hideKeyboard(view: View) {
        val manager = view.context.getSystemService(InputMethodManager::class.java)
        manager?.hideSoftInputFromWindow(view.windowToken, 0)
    }
}

/** Answers the insets the platform reports, which a safe area lays out against. */
object VarnInsets {
    fun of(view: View, density: Float): JSONObject {
        val insets = view.rootWindowInsets
            ?: return JSONObject().put("top", 0).put("right", 0).put("bottom", 0).put("left", 0)

        val bars = insets.getInsets(android.view.WindowInsets.Type.systemBars())

        return JSONObject()
            .put("top", bars.top / density)
            .put("right", bars.right / density)
            .put("bottom", bars.bottom / density)
            .put("left", bars.left / density)
    }
}

File: renderers/ios/VarnActions.swift
import UIKit

/// Performs the imperative action a ref asked for, refusing a name the renderer has no answer for.
enum VarnActions {
    static func perform(_ method: String, on view: UIView, arguments: [String: Any]) throws -> Bool {
        switch method {
        case "focus":
            return view.becomeFirstResponder()

        case "blur":
            return view.resignFirstResponder()

        case "scrollTo":
            guard let scroll = view as? UIScrollView else {
                throw RendererError.missing("scrollTo needs a scrolling view")
            }

            let point = CGPoint(x: VarnValue.number(arguments["x"]) ?? 0, y: VarnValue.number(arguments["y"]) ?? 0)
            scroll.setContentOffset(point, animated: arguments["animated"] as? Bool ?? false)
            return true

        case "play", "pause":
            guard let video = view as? VarnVideoView else {
                throw RendererError.missing("\(method) needs a video")
            }

            if method == "play" {
                video.player.play()
            } else {
                video.player.pause()
            }

            return true

        case "seek":
            guard let seconds = VarnValue.number(arguments["seconds"]).map(Double.init) else {
                throw RendererError.malformed("seek is asked for a number of seconds")
            }

            if let sound = view as? VarnAudioView {
                sound.seek(to: seconds)
                return true
            }

            guard let video = view as? VarnVideoView else {
                throw RendererError.missing("seek needs a sound or a film")
            }

            video.seek(to: seconds)
            return true

        case "capturePhoto", "startRecording", "stopRecording":
            guard let camera = view as? VarnCameraView else {
                throw RendererError.missing("\(method) needs a camera")
            }

            if method == "capturePhoto" {
                camera.capturePhoto()
            } else if method == "startRecording" {
                camera.startRecording()
            } else {
                camera.stopRecording()
            }

            return true

        case "toggleMark", "setLink", "clearLink", "copy", "cut", "paste", "selectAll":
            guard let editor = view as? VarnRichEditor else {
                throw RendererError.missing("\(method) needs an editor")
            }

            return edit(method, on: editor, arguments: arguments)

        default:
            throw RendererError.missing("the renderer has no action named \(method)")
        }
    }

    /// Applies what a toolbar asked of an editor, which is a mark, a link or the clipboard.
    private static func edit(_ method: String, on editor: VarnRichEditor, arguments: [String: Any]) -> Bool {
        editor.becomeFirstResponder()

        switch method {
        case "toggleMark":
            guard let mark = arguments["mark"] as? String else {
                return false
            }

            editor.toggle(mark)

        case "setLink":
            editor.link(arguments["url"] as? String ?? "")

        case "clearLink":
            editor.link(nil)

        case "selectAll":
            editor.selectAll(nil)
            editor.report()

        case "copy":
            editor.copy(nil)

        case "cut":
            editor.cut(nil)

        default:
            editor.paste(nil)
        }

        return true
    }
}
